### PR TITLE
[trainer] add sanity evaluation option 

### LINKED ISSUE #31047: Add sanity validation steps
### Feature request

@RylanSchaeffer

Add support for sanity validation steps before training begins, to ensure evaluation works correctly. For inspiration, Pytorch Lightning has a validation sanity check feature which is similar to what I am requesting.

### Motivation

I would like to have the option to run evaluation before training starts, either for metrics or debugging purposes.

### Your contribution

I can add support for this when given the green light.

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do?
This PR adds the possibility to perform a sanity check for evals before training.
Fixes https://github.com/huggingface/transformers/issues/31047#issuecomment-2137495927 

TODO: add a test

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2723,6 +2726,18 @@ def _issue_warnings_after_load(self, load_result):
                 f"There were unexpected keys in the checkpoint model loaded: {load_result.unexpected_keys}."
             )
 
+    def _evaluate(self, trial, ignore_keys_for_eval):
```
Comment (muellerzr) [inline src/transformers/trainer.py:2729]:
```suggestion
    def _evaluate(self, trial, ignore_keys_for_eval, skip_scheduler=False):
```
We need to disable scheduler skipping, otherwise the scheduler will be skipped on the first step!

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2723,6 +2726,18 @@ def _issue_warnings_after_load(self, load_result):
...
+        metrics = self.evaluate(ignore_keys=ignore_keys_for_eval)
+        self._report_to_hp_search(trial, self.state.global_step, metrics)
+
+        # Run delayed LR scheduler now that metrics are populated
+        if isinstance(self.lr_scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (muellerzr) [inline src/transformers/trainer.py:2734]:
```suggestion
        if isinstance(self.lr_scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau) and not skip_scheduler:
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2175,6 +2175,9 @@ def _inner_training_loop(
         grad_norm: Optional[float] = None
         self.control = self.callback_handler.on_train_begin(args, self.state, self.control)
 
+        if args.sanity_evaluation:
+            self._evaluate(trial, ignore_keys_for_eval)
```
Comment (muellerzr) [inline src/transformers/trainer.py:2179]:
```suggestion
            self._evaluate(trial, ignore_keys_for_eval, True)
```

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -1454,6 +1457,11 @@ class TrainingArguments:
         metadata={"help": "Break eval metrics calculation into batches to save memory."},
     )
 
+    sanity_evaluation: bool = field(
+        default=False,
+        metadata={"help": "Sanity check for the evaluation step."},
```
Comment (muellerzr) [inline src/transformers/training_args.py:1462]:
```suggestion
        metadata={"help": "Whether to run through the entire `evaluation` step at the very beginning of training as a sanity check."},
```

**Review by muellerzr — COMMENTED** [review]
Thanks! Overall a fan of this refactor, however let's be careful not to step the scheduler until we need to 😉

File: src/transformers/trainer.py (hunk 3 above)
Comment (SunMarc) [reply src/transformers/trainer.py:2179]: Done ! but I also added the arg name for readability

**Review by muellerzr — APPROVED** [review]
Thanks! Looks great to me, cc @amyeroberts for final review (post tests I imagine ;) )

**Review by LysandreJik — APPROVED** [review]
